polyfill/index.d.ts

```diff
@@ -185,6 +185,47 @@ export namespace Temporal {
     toString(tzLike?: Temporal.TimeZone | string): string;
   }
 
+  /**
+   * A `Temporal.Calendar` is a representation of a calendar system. It includes
+   * information about how many days are in each year, how many months are in
+   * each year, how many days are in each month, and how to do arithmetic in\
+   * that calendar system.
+   *
+   * See https://tc39.es/proposal-temporal/docs/calendar.html for more details.
+   */
+  export class Calendar {
+    static from(item: Temporal.Calendar | string): Temporal.Calendar;
+    constructor(calendarIdentifier: string);
+    readonly id: string;
+    year(date: Temporal.Date): number;
+    month(date: Temporal.Date): number;
+    day(date: Temporal.Date): number;
+    dayOfWeek(date: Temporal.Date): number;
+    dayOfYear(date: Temporal.Date): number;
+    weekOfYear(date: Temporal.Date): number;
+    daysInMonth(date: Temporal.Date): number;
+    daysInYear(date: Temporal.Date): number;
+    isLeapYear(date: Temporal.Date): boolean;
+    dateFromFields(fields: DateLike, options: AssignmentOptions, constructor: function): Temporal.Date;
+    dateTimeFromFields(fields: DateTimeLike, options: AssignmentOptions, constructor: function): Temporal.DateTime;
+    yearMonthFromFields(fields: YearMonthLike, options: AssignmentOptions, constructor: function): Temporal.YearMonth;
+    monthDayFromFields(fields: MonthDayLike, options: AssignmentOptions, constructor: function): Temporal.MonthDay;
+    plus(
+      date: Temporal.Date,
+      duration: Temporal.Duration,
+      options: ArithmeticOptions,
+      constructor: function
+    ): Temporal.Date;
+    minus(
+      date: Temporal.Date,
+      duration: Temporal.Duration,
+      options: ArithmeticOptions,
+      constructor: function
+    ): Temporal.Date;
```

Comment:
@justingrant I have no idea how to express this in TypeScript: the return type of these methods should be whatever is returned by the passed-in constructor. I suspect what I've written here is technically correct but not specific enough.

For example, if you do
`calendar.dateFromFields(fields, options, Temporal.Date)` then you get a `Temporal.Date`; if you do `class MyDate extends Temporal.Date {}; calendar.dateFromFields(fields, options, MyDate)` then you get a `MyDate`.

Reply:
@ptomato - the magic you're looking for is this utility type: 

```typescript
type ConstructorOf<T> = new (...args: any[]) => T;
```

Translating into English: 
*  `type` - declare a non-exported type...
* `ConstructorOf` - ...called `ConstructorOf...
* `<T>` - ...that takes a generic type parameter named `T`...
* `new` - ... and this type represents a constructor function...
* `(...args: any[])` - ...that accepts any arguments...  
*  `=> T` - and constructs an instance of type `T`

So you'd use `ConstructorOf<Temporal.Date>` instead of `function` for that parameter type. Easy peasy!

What's not so easy is how to get type safety for extra calendar fields like Japanese `era`. What you want is to enable app developers to write code like this: 

```typescript
// in IDE, you'll get autocomplete that includes `era` in these cases
const nowJapanese = Temporal.now.dateTime().withCalendar('japanese');
const era = nowJapanese.era;  // is this supported on the JS side?  Would be cool if so!
const alsoEra = nowJapanese.getFields().era;
const changeTheEra = nowJapanese.with({era: 'someEraName'});

// will give you TS compiler errors
const era = nowJapanese.eeeera;
const alsoEra = nowJapanese.getFields().eeeera;
const changeTheEra = Temporal.now.dateTime().withCalendar('japanese').with({eeeera: 'someEraName'});
```

That's do-able but requires passing generic type parameters down the type hierarchy so that `Date.prototype.getFields` knows its return type, which in turn depends on the calendar, which in turn would depend on a JPN-specific "extra fields" type. One way to implement it is something like below, but I think there's probably a cleaner way to do it.

The complexity below probably isn't needed now, but will be needed for TS users to successfully access those extra fields without ugly type casts. I'd be happy to help you implement the types to make this work. 

If you want, I could help you with this heavy lifting below. The `ConstructorOf`-trick is TS-noob-friendly, but this stuff below is not.

Q: will extra fields added by calendars show up on `Date`, `DateTime`, etc. as property getters?  Or just via `getFields()`?  The code below assumes the former, but either one easily do-able on the TS side once the basic types are all hooked up together. 

```typescript

export class IsoCalendar extends Calendar<{}> {} // no extra fields

export class JapaneseCalendar extends Calendar<{ era: string }> {}

export class Calendar<Extra extends Record<string, any> = {}, Internal extends Record<string, any> = Extra> {
  static from(item: this | string): Temporal.Calendar<Extra>;
  dateFromFields(
    fields: DateLike<Extra, Internal>,
    options: AssignmentOptions,
    constructor: ConstructorOf<Temporal.Date<Extra, Internal>>
  ): Temporal.Date<Extra, Internal>;
. . .
}

export type DateLike<Extra extends Record<string, any>> = {
  year?: number;
  month?: number;
  day?: number;
} & Partial<Extra>; // extends DateLike with extra calendar fields

export class Date<Extra extends Record<string, any> = {}, Internal extends Record<string, any> = Extra> implements Required<DateLike<Extra>> {
  static from(
    item: Temporal.Date<Extra, Internal> | DateLike<Extra> | string,
    options?: AssignmentOptions
  ): Temporal.Date<Extra, Internal>;
  getFields(): Required<DateLike<Extra>>;
. . .
}
```

BTW, I'd suggest not exporting the `*ISOCalendarFields` types. Calendar implementers can get to those types via `Date['getISOCalendarFields']` (that means "type of getISOCalendarFields on the Date type") but by not exporting them they won't show up in autocomplete for noobs. Thats good.

Reply:
Thanks very much for the explanation. I've updated the PR to incorporate your advice.

I have left out the extra fields for the time being, because (1) I haven't wrapped my head around it yet and (2) we are not shipping the initial polyfill with any calendars except ISO 8601 implemented, so there won't be any extra fields yet.

I had to change `any` to `unknown` in the `ConstructorOf` signature and I'm not quite sure why, but the linter complained about "unexpected any, replace with a different type".

To answer your other question,
> will extra fields added by calendars show up on Date, DateTime, etc. as property getters? Or just via getFields()?

The intention is that they do. The discussion about this is in https://github.com/tc39/proposal-temporal/issues/291#issuecomment-585939337, and re-reading it I noticed that it does say that all getters for calendars known to ECMA-402 are predefined, which I had misremembered! So `era` could just be listed in the TypeScript types file. Calendars defined in userland would still be able to add their own fields, though, so I think we still need to have something like what you sketched here.

Reply:
> I had to change `any` to `unknown` in the `ConstructorOf` signature and I'm not quite sure why, but the linter complained about "unexpected any, replace with a different type".

`any` means "I'm too lazy to figure out the type so I'll just opt out of TS in this case" while `unknown` means "I'll defer type checking until I actually have to use the type". The former is generally frowned upon in modern TS code because the latter is more type-safe, hence the linter message. In the case you're using it, it doesn't matter one way or the other since you're not actually type-checking the arguments. 

You probably already know this, but in case not: it's easy to tell the linter to shut up with an [inline comment](https://eslint.org/docs/2.13.1/user-guide/configuring#disabling-rules-with-inline-comments) when there isn't a real problem there.  No shame in disabling lint rules that aren't actually adding value. ;-)

Congrats for getting the first calendar stuff checked in... it's a lot of work for you and the team, so I imagine it's gratifying to get it in. Nice work!

Let me know when you're ready for the "extra fields" case and I'd be happy to work together to figure out the TS part.